### PR TITLE
[clang][fat-lto-objects] Make module flags match non-FatLTO pipelines

### PR DESCRIPTION
**Pull request by ilovepi**
In addition to being rather hard to follow, there isn't a good reason
why FatLTO shouldn't just share the same code for setting module flags
for (Thin)LTO. This patch simplifies the logic and makes sure we use set
these flags in a consistent way, independent of FatLTO.

Additionally, we now test that output in the .llvm.lto section actually
matches the output from Full and Thin LTO compilation.

### CODE REVIEW

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 1]
```diff
@@ -1036,7 +1041,8 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
   if (!actionRequiresCodeGen(Action) && CodeGenOpts.VerifyModule)
     MPM.addPass(VerifierPass());
 
-  if (Action == Backend_EmitBC || Action == Backend_EmitLL) {
+  if (Action == Backend_EmitBC || Action == Backend_EmitLL ||
+      CodeGenOpts.FatLTO) {
```
Comment (teresajohnson) [inline clang/lib/CodeGen/BackendUtil.cpp:1045]: What is the Action type with FatLTO?

Comment (ilovepi) [reply clang/lib/CodeGen/BackendUtil.cpp:1045]: I believe its `Backend_EmitObj` .  I think that gets a little complicated when `-S` or `-emit-llvm` is passed, since I think we then select  `Backend_EmitLL` or `Backend_EmitBC`. https://github.com/llvm/llvm-project/blob/d82e93e7f129d9e8b72570efdf4a15d6ec3d4336/clang/lib/Driver/Driver.cpp#L4770

Comment (teresajohnson) [reply clang/lib/CodeGen/BackendUtil.cpp:1045]: I see, I guess then the check for the CodeGenOpt is needed here to catch this case.

**Review by teresajohnson — APPROVED** [review]

**Comment by ilovepi** [comment]
Hmm, not sure why this comes out differently on windows in the presubmit.

**Comment by ilovepi** [comment]
Seems like "PIE Level" is different, Lets see if passing `-fPIE` will satisfy the windows bot
